controllers/cloud_config_sync_controller.go

```diff
@@ -0,0 +1,196 @@
+	}
+
+	return nil, fmt.Errorf("default key %s does not found in source configmap %s", defaultConfigKey, client.ObjectKeyFromObject(source))
+}
+
+func (r *CloudConfigReconciler) isCloudConfigEqual(source *corev1.ConfigMap, target *corev1.ConfigMap) bool {
+	return source.Immutable == target.Immutable &&
+		reflect.DeepEqual(source.Data, target.Data) && reflect.DeepEqual(source.BinaryData, target.BinaryData)
+}
+
+func (r *CloudConfigReconciler) syncCloudConfigData(ctx context.Context, source *corev1.ConfigMap, target *corev1.ConfigMap) error {
+
+	target.SetName(cloudConfigMapName)
+	target.SetNamespace(r.TargetNamespace)
+	target.Data = source.Data
+	target.BinaryData = source.BinaryData
+	target.Immutable = source.Immutable
+
+	// check if target config exists, crate if not
+	err := r.Get(ctx, client.ObjectKeyFromObject(target), &corev1.ConfigMap{})
+
+	if err != nil && errors.IsNotFound(err) {
+		err = r.Create(ctx, target)
+	} else if err != nil {
+		return err
+	}
+
+	return r.Update(ctx, target)
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *CloudConfigReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	watcher, err := NewObjectWatcher(WatcherOptions{
+		Cache:  mgr.GetCache(),
+		Scheme: mgr.GetScheme(),
+	})
+	if err != nil {
+		return err
+	}
+	r.watcher = watcher
+
+	build := ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.ConfigMap{}, builder.WithPredicates(ownCloudConfigPredicate(r.TargetNamespace))).
+		Watches(
+			&source.Kind{Type: &corev1.ConfigMap{}},
+			&handler.EnqueueRequestForObject{},
+			builder.WithPredicates(cloudConfigMapPredicates(r.InfrastructureResource.Spec.CloudConfig.Name)),
+		)
+
+	return build.Complete(r)
+}
+
+func ownCloudConfigPredicate(targetNamespace string) predicate.Funcs {
+	isOwnCloudConfigMap := func(obj runtime.Object) bool {
+		configMap, ok := obj.(*corev1.ConfigMap)
+		return ok && configMap.GetNamespace() == targetNamespace && configMap.GetName() == cloudConfigMapName
+	}
+
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool { return false },
+		UpdateFunc: func(e event.UpdateEvent) bool { return isOwnCloudConfigMap(e.ObjectNew) },
+		DeleteFunc: func(e event.DeleteEvent) bool { return isOwnCloudConfigMap(e.Object) },
+		// filter out generic events for avoid redundant reconciliation and be sure that cloud config changed if event was emitted
+		GenericFunc: func(e event.GenericEvent) bool { return false },
+	}
+}
+
+func cloudConfigMapPredicates(infraCloudConfigName string) predicate.Funcs {
+	isCloudConfigMap := func(obj runtime.Object) bool {
+		configMap, ok := obj.(*corev1.ConfigMap)
+
+		if ok {
+			isCloudConfName := configMap.GetName() == infraCloudConfigName || configMap.GetName() == managedCloudConfigMapName
+			isCloudConfNamespace := configMap.GetNamespace() == openshiftManagedConfigNamespace || configMap.GetNamespace() == openshiftConfigNamespace
+
+			return isCloudConfName && isCloudConfNamespace
+		}
+
+		return ok
```

Comment:
I would reverse the if statement here to reduce indentation
```suggestion
		configMap, ok := obj.(*corev1.ConfigMap)
		if !ok {
			return false
		}
		
		isCloudConfName := configMap.GetName() == infraCloudConfigName || configMap.GetName() == managedCloudConfigMapName
		isCloudConfNamespace := configMap.GetNamespace() == openshiftManagedConfigNamespace || configMap.GetNamespace() == openshiftConfigNamespace

		return isCloudConfName && isCloudConfNamespace
```